### PR TITLE
fix: services.<name>.service.build.context

### LINKED ISSUE #208: services.<name>.service.build.context doesn't work
When setting the option, the resulting `docker-compose.yaml` always has an image set by `services.<name>.service.image`.

These lines seem to be the cause https://github.com/hercules-ci/arion/blob/f0436c8478d1b5530c115a6b1202c4478dfe2f81/src/nix/modules/service/image.nix#L167-L177

`service.image` ends up **always** being set which also always sets `nixBuild` - at least that's my evaluation. I tried to make it optional, but failed to set the `service.build....` fields :shrug: 

```diff
diff --git a/src/nix/modules/service/docker-compose-service.nix b/src/nix/modules/service/docker-compose-service.nix
index 061ae8b..c038ac6 100644
--- a/src/nix/modules/service/docker-compose-service.nix
+++ b/src/nix/modules/service/docker-compose-service.nix
@@ -86,7 +86,8 @@ in
       description = serviceRef "environment";
     };
     service.image = mkOption {
-      type = str;
+      type = nullOr str;
+      default = null;
       description = serviceRef "image";
     };
     service.command = mkOption {
@@ -328,8 +329,9 @@ in
       volumes
       environment
       sysctls
-      image
       ;
+  } // lib.optionalAttrs (config.service.image != null) {
+    inherit (config.service) image;
   } // lib.optionalAttrs (config.service.build.context != null) {
     inherit (config.service) build;
   } // lib.optionalAttrs (cap_add != []) {
diff --git a/src/nix/modules/service/image.nix b/src/nix/modules/service/image.nix
index 706a5c1..b6aa289 100644
--- a/src/nix/modules/service/image.nix
+++ b/src/nix/modules/service/image.nix
@@ -163,17 +163,31 @@ in
       '';
     };
   };
-  config = {
-    build.image = builtImage;
-    build.imageName = config.build.image.imageName;
-    build.imageTag =
+  config = let
+
+    image = builtImage;
+    imageName = config.build.image.imageName;
+    imageTag =
                  if config.build.image.imageTag != ""
                  then config.build.image.imageTag
                  else lib.head (lib.strings.splitString "-" (baseNameOf config.build.image.outPath));
 
-    service.image = lib.mkDefault "${config.build.imageName}:${config.build.imageTag}";
-    image.rawConfig.Cmd = config.image.command;
+    serviceImage = lib.mkDefault "${imageName}:${imageTag}";
+    hasContext = config.service.build.context != null;
 
-    image.nixBuild = lib.mkDefault (priorityIsDefault options.service.image);
-  };
+  in lib.mkMerge [
+    {
+      build = {
+        inherit image imageName imageTag;
+      };
+      image.rawConfig.Cmd = config.image.command;
+      image.nixBuild = false;
+    }
+    ( lib.mkIf (hasContext == false)
+    {
+      service.image = serviceImage;
+      image.nixBuild = lib.mkDefault (priorityIsDefault options.service.image);
+    })
+  ];
 }

```

Maybe somebody with more `nix` and `nixpkgs` knowledge can help solve this.

### PR DESCRIPTION
**Pull request by LoveIsGrief**
- [x] Support services.<name>.service.build.context see 638c4b8  for more details
 - [x] Add test

Closes #208

### CODE REVIEW

**Comment by roberth** [comment]
bors try

**Comment by LoveIsGrief** [comment]
bors try

**Comment by roberth** [comment]
bors try

Gotta do something about bors. That reviewer link doesn't work the way I want it to.

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,5 @@ dist/
 dist-newstyle/
 cabal.project.local
 
+*.swp
```
Comment (roberth) [inline .gitignore:8]: Would recommend to add it to your personal user gitignore, but I don't mind having a copy here.

File: src/haskell/test/Arion/NixSpec.hs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE OverloadedStrings #-}
 module Arion.NixSpec
   ( spec
+    , contextSpec
```
Comment (roberth) [inline src/haskell/test/Arion/NixSpec.hs:4]: Could you change `spec` to look like the following?

```haskell
spec = describe "evaluateComposition" $ do
  it "matches an example" $ do
    ...
  it "matches a build.context example" $ do
    ...

```

`describe` forms a tree structure, where `it` are the leaves, so we can make use of that here, keep the interface simple.

Comment (LoveIsGrief) [reply src/haskell/test/Arion/NixSpec.hs]: Done in  3588b01

File: .gitignore (hunk 1 above)
Comment (LoveIsGrief) [reply .gitignore:8]: Cheers. I haven't set that up yet. Still setting up home-manager.

**Review by roberth — APPROVED** [review]
Looks good!

**Comment by roberth** [comment]
Awesome, thanks @LoveIsGrief !

**Comment by LoveIsGrief** [comment]
Thank you for the quick response and for arion @roberth !

**Pull request merged**